Add Scheduled Publish functionality (Lombiq Technologies: OCORE-16)

Piedone opened the pull request:
I.e. the Publish Later module, see #3991. This is using a simple logic to schedule when a content item should be published, not a whole schedule task infrastructure. Mostly follows the 1.x. Publish Later module's behavior.

Piedone commented:
@Skrypt in the meeting you said:

>     Ok we can't merge this PublishLater PR if you are not using the ContentManager event as it won't generate a Lucene Index document
>     Ok I see it now
> awaitserviceProvider.GetRequiredService<IContentManager>().PublishAsync(item);
> https://github.com/OrchardCMS/OrchardCore/pull/5639/files#diff-3d65de6074032bb7e3cf0c0f9822aaaf
> The PublishLaterPartIndexProvider mioght require a boolean
>     To publish or unpublish
>     And probably something to store the contentItemID
>     As you could hit that button many times and create multiple drafts
>     So it should only store the last draft in that index  table

1. Where would I use ContentManager events? The code is doing a standard publish.
2. What do you mean the `PublishLaterPartIndexProvider` might requrire a boolean to publish or unpublish? Are you talking about extending this with archive later functionality?
3. What scenario exactly are you concerned regarding hitting the button many times, when would multiple drafts be created? I don't see how this can happen (the Publish Later button is just a Save Draft button in disguise).

Skrypt commented:
1. It's fine like I said. It's calling the Publish events in the DefaultContentManager.
2. Up to you to decide on the design. Maybe a different Part could be used for the ArchiveLater with a different background task process for these.
3. Needs to be tested to see what happens. Let's say I'm using publish later button once. Then I decide to use the publish button, then I decide to reuse the publish later button. What will happen? I mean, we need to test some use cases like that to make sure that the PublishLaterPartIndexProvider stores enough informations to make sure it doesn't try to publish 2 different drafts or something like it. As far as I remember the indexing providers can store a published and a drafted version of the same content item. Need to see if everything works fine still when that happens.

Piedone commented:
1. OK then!
2. OK, I was just confused on what exactly are you referring to, whether it was something concerning the current feature set or an Archive Later extension.
3. I've played around and while there is an index record (potentially with a `ScheduledPublishUtc` value) saved for every version this doesn't seem to be an issue. To be sure I've added a constraint on versions to the query in question. For the case of when somebody schedules a publish but then also publishes manually (making scheduled publish unneeded) I added a clearing handler.

Is it possible to have an IndexProvider that only creates index records for the latest version of the content item (or just one per content item in general)? I think a check on `contentItem.Published` and `contentItem.Latest` in `Map()` should work similarly to how a couple of other providers do. Just so there won't be unnecessary index records.

Skrypt commented:
In the Map yes.

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.PublishLater/Indexes/PublishLaterPartIndex.cs:
```diff
@@ -0,0 +1,33 @@
+                .Map(contentItem =>
+                {
+                    var publishLaterPart = contentItem.As<PublishLaterPart>();
+                    if (publishLaterPart != null)
```
as @Skrypt says something like `|| !(contentItem.Published || contentItem.Latest) return null` (not sure of the exact logic you require) will stop this creating indexes for every item.

Piedone replied:
Thanks, done that. Now there will be only an index record for the latest draft, and only if it actually has a date scheduled.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.PublishLater/Views/PublishLaterPart.SummaryAdmin.cshtml:
```diff
@@ -0,0 +1,8 @@
+    <div>
+        <span class="hint">@T["Scheduled to be published on {0}", (object)(await DisplayAsync(await New.DateTime(Utc: Model.ScheduledPublishUtc)))]</span>
+    </div>
+}
```
newline

Skrypt commented on src/OrchardCore.Modules/OrchardCore.PublishLater/Views/_ViewImports.cshtml:
```diff
@@ -0,0 +1,8 @@
+
+@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
+@addTagHelper *, OrchardCore.DisplayManagement
+@addTagHelper *, OrchardCore.ResourceManagement
```
newline

Piedone replied:
Can you elaborate?

Skrypt replied:
Every time you see a little circle icon in the end of your files that means it requires you to add a new line at the end for Git.

Skrypt replied:
This icon : 
![image](https://user-images.githubusercontent.com/3228637/75927142-c89b7700-5e39-11ea-96c2-3ef984c9c887.png)

hishamco replied:
I think it's by default added whenever you are using a VS

Piedone replied:
Ah OK, that's interesting, thank for clarifying. Reading about it seems to be a leftover from the 70s :D. I fixed these.

Piedone commented:
Any other suggestions/remarks or can we merge this?

agriffard commented:
Waiting for review and approval.

/cc @Skrypt @hishamco @deanmarcussen

hishamco commented on src/OrchardCore.Modules/OrchardCore.PublishLater/Startup.cs:
```diff
@@ -0,0 +1,37 @@
+using Fluid;
```
Fluid should be go after `Microsoft.*` namespace

Piedone replied:
Why? I'm using [this VS extension](https://marketplace.visualstudio.com/items?itemName=VisualStudioPlatformTeam.ProductivityPowerPack2017) BTW that removes unneeded usings on save (small code cleanliness improvement), as well as sorts them alphabetically (not that the order would matter here much but if we want a specific order then that one is simple and automated, and you don't have to think about it, which is good for something that doesn't matter in the end :)).

jtkech replied:
We have an `.editorconfig` file, the only rule is to put the System ones first, the other ones being ordered alphabetically.

    dotnet_sort_system_directives_first = true

So Fluid "should" be before Microsoft.* ;)

Piedone replied:
Hmm, OK, the editorconfig is supposed to be applied, though no warning, message or anything is visible for this file. With all other namespaces the system ones are actually at the top so I don't know what's the difference with Fluid. Anyway, moved it to the top by hand.

jtkech replied:
> the editorconfig is supposed to be applied

Under the VStudio editor you still need to right click and do `Remove and Sort Usings`.

Piedone replied:
Yeah this is what the linked extension also does on save (it actually does alphabetic sorting just by default and follows whats's in the editorconfig otherwise). It seems Fluid is somehow breaking this feature.

hishamco commented on src/OrchardCore.Modules/OrchardCore.PublishLater/Views/PublishLaterPart.Edit.cshtml:
```diff
@@ -0,0 +1,18 @@
+{
+    <div class="btn-group">
+        <input asp-for="ScheduledPublishLocalDateTime" type="datetime-local" class="form-control content-preview-select" />
+        <button class="btn btn-primary" type="submit" name="submit.Save" value="submit.PublishLater">@T["Publish&nbsp;Later"]</button>
```
I'm not sure why we can't use `@T["Publish Later"]` instead?!! I don't think we need to use `&nbsp;`

Piedone replied:
There's a nbsp here and in the cancel button because otherwise the buttons' texts can be wrapped on certain screen sizes. E.g. this is with nbsp:

![image](https://user-images.githubusercontent.com/1976647/79048177-870da100-7c1b-11ea-82d8-fb7676aadbb0.png)

This is without:
![image](https://user-images.githubusercontent.com/1976647/79048193-98ef4400-7c1b-11ea-8f13-7e7d15a83b08.png)

Probably a matter of taste but I like the first one better.

Piedone replied:
Hmm but now that I think about it this would indeed be nicer from CSS, you're right. I'll change this.

jtkech replied:
At least it proves that when using an `IViewLocalizer` the main string is not html encoded ;) Only `{0}` parameters are encoded, this allows to include html tags, an entity as `&nbsp;` ...

So the main strings of our translations using `T[...]` may need to be html pre-encoded, not sure we are aware of this ;)

Piedone replied:
Changed this and the cancel button to do the same from CSS.

hishamco commented on src/OrchardCore.Modules/OrchardCore.PublishLater/Views/PublishLaterPart.Edit.cshtml:
```diff
@@ -0,0 +1,18 @@
+    @if (Model.ScheduledPublishUtc.HasValue)
+    {
+        <div class="btn-group">
+            <button class="btn btn-primary" type="submit" name="submit.Save" value="submit.CancelPublishLater">@T["Cancel&nbsp;Publish&nbsp;Later"]</button>
```
Same here

Piedone commented on src/OrchardCore.Modules/OrchardCore.PublishLater/Drivers/PublishLaterPartDisplayDriver.cs:
```diff
@@ -0,0 +1,80 @@
+{
+    public class PublishLaterPartDisplayDriver : ContentPartDisplayDriver<PublishLaterPart>
+    {
+        private readonly IHttpContextAccessor _hca;
```
Sure. In O1 we commonly used this abbreviation so I'd suggest making it popular here too.

hishamco replied:
It's unreadable, furthermore we already used readable names overall the code

Piedone commented on src/OrchardCore.Modules/OrchardCore.PublishLater/Manifest.cs:
```diff
@@ -0,0 +1,11 @@
+[assembly: Module(
+    Name = "Publish Later",
+    Author = "The Orchard Team",
+    Website = "https://orchardproject.net",
```
We could have a const for both.

agriffard replied:
ok constants available @Piedone : https://github.com/OrchardCMS/OrchardCore/pull/5905/commits/aff21dcaeb54373ca0bc28b3ab7d6a1c4bedd6be

```
    Author = ManifestConstants.OrchardCoreTeam,
    Website = ManifestConstants.OrchardCoreWebsite,
    Version = ManifestConstants.OrchardCoreVersion,
```

Piedone replied:
Great, thanks! If it gets into dev earlier then this then I'll merge.

agriffard commented:
If integrated in RC2, we will need to wait that #5905 is merged and then merge dev into it and use the Manifest constants.

agriffard commented:
#5905 merged.
Please merge dev and use the Manifest constants.

Piedone commented:
Done.
